File: koko-eating-bananas/koko-eating-bananas.cpp
class Solution {
    bool solve(vector<int>&nums,int mid,int h){
        int min_hours=0;
        
        for(int i=0;i<nums.size();i++){
            if(nums[i]%mid==0)
                min_hours+=nums[i]/mid;
            else
                min_hours+=nums[i]/mid+1;
        }
        if(min_hours>h)
                return false;
        return true;
    }
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        
        int lo=0,hi=1e9+1e5;
        long long sum=0;
        for(int i=0;i<piles.size();i++){
            sum+=piles[i];
        }
        if(sum<=h)
            return 1;
        while(lo<hi){
            int mid=lo+(hi-lo)/2;
            if(solve( piles,mid,h))
                hi=mid;
            else
                lo=mid+1;
        }
        return lo;
    }
};

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l = 1
        r = max(piles)
        ans = 0
        def solve(m):
            count = 0 
            for p in piles:
                if p<=m:
                    count+=1
                else:
                    count += p//m
                    if p%m!=0:
                        count += 1
            return count<=h
        while l<=r:
            mid = (l+r)//2
            if(solve(mid)):
                r = mid - 1
                ans = mid  
            else:
                l = mid + 1
        return ans
